### PR TITLE
JavaScript II Project Submission

### PR DESCRIPTION
**Pull request by lafriedel**
Completed project goals and attempted some stretch @m-j-hacker

### CODE REVIEW

**Comment by m-j-hacker** [comment]
Thanks for the PR

File: assignments/array-methods.js [hunk 1]
```diff
@@ -56,28 +56,84 @@ const runners = [{"id":1,"first_name":"Charmain","last_name":"Seiler","email":"c
 // ==== Challenge 1: Use .forEach() ====
 // The event director needs both the first and last names of each runner for their running bibs.  Combine both the first and last names into a new array called fullName. 
 let fullName = [];
+
+runners.forEach((runner) => {
+    fullName.push(`${runner.first_name} ${runner.last_name}`);
+})
+
```
Comment (m-j-hacker) [inline assignments/array-methods.js:63]: :heavy_check_mark:

File: assignments/array-methods.js [hunk 2]
```diff
@@ -56,28 +56,84 @@ const runners = [{"id":1,"first_name":"Charmain","last_name":"Seiler","email":"c
...
+// Alternative solution just for fun
+
+const diffSize = runners.filter((runner) => {
+    if (runner.shirt_size === "L") return runner;
+}).map((runner) => {
+    return `${runner.first_name} ${runner.last_name}`;
+})
+
+console.log(diffSize);
+
```
Comment (m-j-hacker) [inline assignments/array-methods.js:99]: Good use of array methods! :1st_place_medal:

File: assignments/array-methods.js [hunk 3]
```diff
@@ -56,28 +56,84 @@ const runners = [{"id":1,"first_name":"Charmain","last_name":"Seiler","email":"c
...
+let deductionReceipts = [];
 
-// Problem 2
+runners.forEach((company) => {
+    deductionReceipts.push({company: `${company.company_name}`, amount: `$${company.donation}.00`});
+})
 
-// Problem 3
\ No newline at end of file
+console.log(deductionReceipts);
```
Comment (m-j-hacker) [inline assignments/array-methods.js:139]: Excellent work! :1st_place_medal:

File: assignments/callbacks.js [hunk 4]
```diff
@@ -24,29 +24,88 @@ const items = ['Pencil', 'Notebook', 'yo-yo', 'Gum'];
...
 }
 
+getLength(items, function(length) {
+  console.log(length);
+});
+
+
 function last(arr, cb) {
   // last passes the last item of the array into the callback.
+  return cb(arr.pop());
```
Comment (m-j-hacker) [inline assignments/callbacks.js:37]: Nice use of `.pop()`!

**Review by m-j-hacker — COMMENTED** [review]
# JavaScript - II

With some basic JavaScript principles in hand, we can now expand our skills out even further by exploring callback functions, array methods, and closure.  Finish each task in order as the concepts build on one another.

## Task 1: Set Up The Project With Git

* [x] Fork the project into your GitHub user account
* [x] Clone the forked project into a directory on your machine
* [x] You are now ready to build this project with your preferred IDE
* [x] To test your `console.log()` statements, open up the index.html file found in the assignments folder and use the developer tools to view the console.  


## Task 2: Callbacks

This task focuses on getting practice with callback functions by giving you an array of values and instructions on what to do with that array.

* [x] Review the contents of the [callbacks.js](assignments/callbacks.js) file.  Notice you are given an array at the top of the page.  Use that array to aid you with your callback functions.

* [x] Write out each function using the `ES5`  `function` keyword syntax.

* [x] Solve the problems listed.  Save the stretch problems until you have completed Tasks 1-4.

## Task 3: Array Methods

Use `.forEach()`, `.map()`, `.filter()`, and `.reduce()` to loop over an array with 50 objects in it. The [array-methods.js](assignments/array-methods.js) file contains several challenges built around a fundraising 5K fun run event.

* [x] Review the contents of the [array-methods.js](assignments/array-methods.js) file.  

* [x] Complete the problems provided to you

* [x] Notice the last three problems are up to you to create and solve.  This is an awesome opportunity for you to push your critical thinking about array methods, have fun with it.

## Task 4: Closures

We have learned that closures allow us to access values in scope that have already been invoked.  

**Hint: Utilize debugger statements in your code in combination with your developer tools to easily identify closure values.**

* [x] Review the contents of the [closure.js](assignments/closure.js) file.  
* [x] Solve the problems listed.  Save the stretch problems until you have completed Tasks 1-4.
* [x] Once you have completed this task please submit a pull request against the original fork.  

## Stretch Goals

* [x] Arrow Function Syntax - [Check out this awesome guide for ES6 arrow syntax](https://medium.freecodecamp.org/when-and-why-you-should-use-es6-arrow-functions-and-when-you-shouldnt-3d851d7f0b26). You will see more and more arrow functions as you progress deeper into JavaScript. Use the [stretch-function-conversion.js](assignments/function-conversion.js) file as a helper challenge to showcase some of the differences between ES5 and ES6 syntax. 

* [ ] Look up what an IIFE is in JavaScript and experiment with them

# Project Manager Feedback
## Great
Excellent work on the assignment, Leigh-Ann! Your code looks great and you completed the objectives and stretch goals! :1st_place_medal: 
## Requested Improvements
I don't have anything to add here for this assignment.
## Questions
* Having used them and shown proficiency, do you find yourself preferring arrow functions over declarative ones?
## Rating {3}
:star: :star: :star:

File: assignments/array-methods.js (hunk 2 above)
Comment (lafriedel) [reply assignments/array-methods.js:99]: Thanks! Before coming to LS I was intimidated by these methods - now chaining makes sense to me. :)

File: assignments/array-methods.js (hunk 3 above)
Comment (lafriedel) [reply assignments/array-methods.js:139]: 👍it was fun to think of different situations

**Comment by lafriedel** [comment]
Definitely liking arrow functions, but I did read today about when NOT to use them (which applies very much to the material this week), so I have to make sure I'm aware of that so I avoid issues.

**Pull request closed without merging**